src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,456 @@
+
+ [Flutter `compute()` function]: https://docs.flutter.dev/cookbook/networking/background-parsing#4-move-this-work-to-a-separate-isolate
+
+
+### Implementing a simple worker isolate
+
+This section shows the implementation for a
+main isolate and the simple worker isolate that it spawns.
+The worker isolate executes a function and then exits,
+sending the main isolate a single message as it exits.
+(The [Flutter `compute()` function][] works in a similar way.)
+
+This example uses the following isolate-related API:
+
+* [`Isolate.spawn()`][] and [`Isolate.exit()`][]
+* [`ReceivePort`][] and [`SendPort`][]
+
+[`Isolate.exit()`]: https://api.dart.dev/stable/dart-isolate/Isolate/exit.html
+[`Isolate.spawn()`]: https://api.dart.dev/stable/dart-isolate/Isolate/spawn.html
+[`ReceivePort`]: https://api.dart.dev/stable/dart-isolate/ReceivePort-class.html
+[`SendPort`]: https://api.dart.dev/stable/dart-isolate/SendPort-class.html
+
+Here’s the code for the main isolate:
+
+```dart
+Future<void> main() async {
+  // Read some data.
+  final jsonData = await _parseInBackground();
+
+  // Use that data
+  print('number of JSON keys = ${jsonData.length}');
+}
+
+// Spawns an isolate and waits for the first message
+Future<Map<String, dynamic>> _parseInBackground() async {
+  final p = ReceivePort();
+  await Isolate.spawn(_readAndParseJson, p.sendPort);
+  return await p.first;
+}
+```
+
+The `_parseInBackground()` function contains the code that
+_spawns_ (creates and starts) the isolate for the background worker,
+and then returns the result:
+
+1. Before spawning the isolate, the code creates a `ReceivePort`,
+   which enables the worker isolate
+   to send messages to the main isolate.
+
+2. Next is the call to `Isolate.spawn()`,
+   which creates and starts the isolate for the background worker.
+   The first argument to `Isolate.spawn()` is the function that
+   the worker isolate executes: `_readAndParseJson`.
+   The second argument is the `SendPort`
+   that the worker isolate can use to send messages to the main isolate.
+   The code doesn't _create_ a `SendPort`;
+   it uses the `sendPort` property of the `ReceivePort`.
+
+3. Once the isolate is spawned, the main isolate waits for the result.
+   Because the `ReceivePort` class implements `Stream`,
+   the [`first`][] property is an easy way to get
+   the single message that the worker isolate sends.
+
+[`first`]: https://api.dart.dev/stable/dart-async/Stream/first.html
+
+The spawned isolate executes the following code:
+
+```dart
+Future _readAndParseJson(SendPort p) async {
+  final fileData = await File(filename).readAsString();
+  final jsonData = jsonDecode(fileData);
+  Isolate.exit(p, jsonData);
+}
+```
+
+The relevant statement is the last one, which exits the isolate,
+sending `jsonData` to the passed-in `SendPort`.
+Message passing between isolates normally involves data copying,
+and thus can be slow and increases with the size of the message —
+O(n) in big O notation.
```

Comment:
Link to https://en.wikipedia.org/wiki/Big_O_notation

Reply:
I wonder is it even useful to talk about Big O notation here? If someone doesn't know it, I'm not sure they're likely to go read a Wikipedia article to understand it. Perhaps a slightly expanded explanation of how the run time scales linearly with the message size would be more useful. Similar for the reference to constant time below.

Reply:
Speaking as someone who knew about big O notation but doesn't use it often, the page was helpful to me. :) It might be good to go into more detail on this page (or in a future page about performance), but for now big O is a handy shorthand.